compiler/rustc_ty_utils/src/implied_bounds.rs

```diff
@@ -56,6 +67,38 @@ fn assumed_wf_types<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx ty::List<Ty
         | DefKind::LifetimeParam
         | DefKind::GlobalAsm
         | DefKind::Closure
-        | DefKind::Generator => ty::List::empty(),
+        | DefKind::Generator => ty::EarlyBinder(ty::Binder::dummy(ty::List::empty())),
     }
 }
+
+fn from_fn_sig<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+) -> ty::Binder<'tcx, &'tcx ty::List<Ty<'tcx>>> {
+    // FIXME(#84533): We probably shouldn't use output types in implied bounds.
+    // This would reject this fn `fn f<'a, 'b>() -> &'a &'b () { .. }`.
+    let unnormalized = tcx.fn_sig(def_id).map_bound(|sig| sig.inputs_and_output);
+    let normalized = normalize_ignoring_obligations(tcx, tcx.param_env(def_id), unnormalized)
+        .unwrap_or_else(|_| ty::Binder::dummy(ty::List::empty()));
+
+    // FIXME(#105948): Use unnormalized types for implied bounds as well.
+    normalized
+}
+
+fn normalize_ignoring_obligations<'tcx, T: TypeFoldable<'tcx>>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    value: T,
+) -> Result<T, NoSolution> {
+    use rustc_infer::infer::TyCtxtInferExt as _;
+    use rustc_infer::traits::ObligationCause;
+    use rustc_trait_selection::traits::query::normalize::QueryNormalizeExt as _;
+
+    let infcx = tcx.infer_ctxt().build();
+    let normalized = infcx
+        .at(&ObligationCause::dummy(), param_env)
+        .query_normalize(value)
```

Comment:
`query_normalize` eagerly evaluates constants causing cycle errors if there are unevaluated constants in the fn sig.

During analysis `normalize` should be used instead.